src/Workspaces/Core/Portable/Formatting/Engine/AbstractFormatEngine.cs

```diff
@@ -271,36 +271,41 @@ void beginningOfTreeTriviaInfoApplier(int i, TriviaData info)
                 return;
             }
 
-            void endOfTreeTriviaInfoApplier(int i, TriviaData info)
+            void endOfTreeTriviaInfoApplier(int i, TokenStream ts, TriviaData info)
             {
-                tokenStream.ApplyEndOfTreeChange(info);
+                ts.ApplyEndOfTreeChange(info);
             }
 
             // remove all trailing indentation
             var triviaInfo = tokenStream.GetTriviaDataAtEndOfTree().WithIndentation(0, context, _formattingRules, cancellationToken);
 
-            triviaInfo.Format(context, _formattingRules, endOfTreeTriviaInfoApplier, cancellationToken);
+            triviaInfo.Format(context, _formattingRules, endOfTreeTriviaInfoApplier, tokenStream, cancellationToken);
         }
 
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/30819", AllowCaptures = false)]
         private void ApplyTriviaOperations(FormattingContext context, TokenStream tokenStream, CancellationToken cancellationToken)
         {
-            // trivia formatting result appliers
-            void regularApplier(int tokenPairIndex, TriviaData info)
+            void regularApplier(int tokenPairIndex, TriviaData info, TokenStream ts)
```

Comment:
order of parameter seems in wrong order?

Reply:
so local function, when given to delegate, it won't allocate like instance method?

Reply:
> order of parameter seems in wrong order?

Fixed in 0dab2852ef86f7589971247267a24ef155148624

> so local function, when given to delegate, it won't allocate like instance method? as long as there is no capture?

Covered by https://github.com/dotnet/roslyn/pull/33291#discussion_r255681619